proposals/js-connection-manager.md

```diff
@@ -0,0 +1,125 @@
+#### Background &amp; intent
+_Describe the desired state of the world after this project? Why does that matter?_
+<!--
+Outline the status quo, including any relevant context on the problem you’re seeing that this project should solve. Wherever possible, include pains or problems that you’ve seen users experience to help motivate why solving this problem works towards top-line objectives. 
+-->
+
+With the existing protocols in libp2p, as well as IPFS subsystems built on top of libp2p, such as Pubsub and the DHT, the need for a connection manager overhaul becomes an import work stream, so that these protocols operate as expected by the users, i.e. out of the box solution.
+
+Currently, js nodes have a reactive connection manager that can be decoupled into two parts: establishing new connections and trimming connections. The former relies on an `autoDial` approach, where each time a new node is discovered it will try to establish a connection with that peer. The later consists on blindly trimming connections when reaching a configurable threshold.
+
+From the current solution, there is a lot of space for improvement with tremendous value for the users. Either evolving the current connection manager to the state of the go-implementation or implementing a fully fledged [Connection Manager v2](https://github.com/libp2p/specs/pull/161) (+ [more notes](https://github.com/libp2p/notes/issues/13)).
+
+With the connection manager overhaul in JS we aim to:
+- Find our closest peers on the network, and attempt to stay connected to n to them
+- Finding, connecting to and protecting our gossipsub peers (same topics search)
+- Finding and binding to relays with AutoRelay
+- Finding and binding to application protocol peers (as needed via MulticodecTopology) -- We should clarify what libp2p will handle intrinsically and what users need to do
+
+All the above scenarios need proper setups from our users at the moment, which means they need to manually find the nodes they want to connect and connect to them, as libp2p connection manager will not be actively looking at meaningful connection to establish.
+
+An active connection manager will enable out of the box creation of meaningful overlay networks, as well as enable browser nodes to bind to AutoRelay nodes, so that they can be dialed from other nodes in the network. Moreover, JS nodes keep the discovered peers stored in the PeerStore. On subsequent node starts, these nodes should prioritize their previous established topology over the bootstrap nodes if possible.
+
+#### Assumptions &amp; hypotheses
+_What must be true for this project to matter?_
+<!--(bullet list)-->
+
+- Web3 developers want to have a reliable pubsub topology out of the box without relying on star servers
+- Web3 developers want to find and connect to other peers in a given scope
+- Browser developers want to have their nodes reachable via more transports than `webrtcSTAR` out of the box
+- PL wants to decrease the load on bootstrap nodes
```

Comment:
We should clarify this a bit. The goal here is that once a node performs an initial query to discover its closest peers, it should no longer need to rely on the Bootstrap nodes unless all other nodes it knows go offline. They're of little value outside of joining the network and initial discovery (unless they happen to be a closest peer to us).

```suggestion
- Reduce the reliance nodes have on Bootstrap nodes outside of joining the network, to reduce the load on those servers
```